src/Build/Globbing/GlobVisitor.cs

```diff
@@ -0,0 +1,63 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//-----------------------------------------------------------------------
+
+
+using System.Collections.Immutable;
+
+namespace Microsoft.Build.Globbing
+{
+    internal abstract class GlobVisitor
+    {
+        public void Visit(IMsBuildGlob glob)
+        {
+            var msbuildGlob = glob as MsBuildGlob;
+            if (msbuildGlob != null)
+            {
+                VisitMSBuildGlob(msbuildGlob);
+            }
+
+            var compositGlob = glob as CompositeGlob;
+            if (compositGlob != null)
+            {
+                VisitCompositeGlob(compositGlob);
+
+                foreach (var globPart in compositGlob.Globs)
+                {
+                    Visit(globPart);
+                }
+            }
+
+            var globWithGaps = glob as MsBuildGlobWithGaps;
+            if (globWithGaps != null)
+            {
+                VisitGlobWithGaps(globWithGaps);
+
+                Visit(globWithGaps.MainGlob);
+            }
+        }
+
+        protected virtual void VisitGlobWithGaps(MsBuildGlobWithGaps globWithGaps)
+        {
+        }
+
+        protected virtual void VisitCompositeGlob(CompositeGlob compositGlob)
+        {
+        }
+
+        protected virtual void VisitMSBuildGlob(MsBuildGlob msbuildGlob)
+        {
+        }
+    }
+
+    internal class ParsedGlobCollector : GlobVisitor
+    {
+        private readonly ImmutableList<MsBuildGlob>.Builder _collectedGlobs = ImmutableList.CreateBuilder<MsBuildGlob>();
```

Comment:
Is it worth keeping a `Builder` around? I thought the idea behind the Immutable* collections was that you could use them in the Haskell way and be ok, and Builders were just an escape hatch in case you had to do some intense one-time manipulation?